ksql-rest-app/src/main/java/io/confluent/ksql/rest/entity/SchemaInfo.java

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.ksql.rest.entity;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Objects;
+
+public class SchemaInfo {
```

Comment:
is there a reason we need a new way of serializing schema information? Why not reuse the existing serde for `Schema`, which we already use internally and also expose as part of the websockets info here https://github.com/confluentinc/ksql/blob/4e4b9ac4101723c209a0f28294998c5c4394a2de/ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/WebSocketSubscriber.java#L101

Reply:
AFAIK Schema isn't set up to be serialized/deserialized. The default serialization may give you something sane, but can jackson deserialize it back into a Schema? One workaround could be to convert it into a generic object first (via serialize/deserialize) and have a generic object for the schema in the API entities.

Compatibility - is the class that implements Schema guaranteed not to change in a way that breaks the API? I think the default serialization leaks implementation-specific fields that could change.

Schema may include some fields that don't make sense to KSQL clients.

Reply:
I thought that's what https://github.com/confluentinc/ksql/blob/master/ksql-rest-app/src/main/java/io/confluent/ksql/rest/entity/SchemaMapper.java was for. Any adjustments to the format can probably be made there?

Reply:
I didn't even know that existed =P

To be fair I don't think its being used for anything other than the web-socket endpoint at the moment. We don't return Schema objects in the API anywhere else.

Let me see what it's doing and jf jt makes sense to return the schema that way.

Reply:
It predates the ws endpoint so it must be used elsewhere. It could get serialized as part of a subfield somewhere.

Reply:
It's not used in any other schema field. My guess is that we had Schema as a field in some earlier iteration of the API and then removed it.

It looks like its using the JsonConverter to convert the Schema to Connect's internal json representation. So this should at least be compatible across versions since connect writes schemas out to kafka in this format. I still don't think its the right approach for the following reasons:

The serialized type names don't match the KSQL type names. We should return the KSQL type names in the schema since atm the fact that we use Connect's type internally is an implementation detail.

The serialized schema includes connect-schema fields that ksql clients wouldn't know what to make of (e.g. "optional")

Those problems can still be solved via the mapper. But I don't think it makes sense to have connect Schema objects in the API client in the first place, since that's a user-facing interface.

Reply:
I agree we don't have to follow the connect schema format if we feel there is a need to have the KSQL one evolve independently. Are you still planning to use the mapper for that though? Seems cleaner than adding yet more wrapper objects.

Reply:
If we use the mapper then we have to have an instance of the connect Schema in the response returned by KafkaRestClient, which I'm hesitant to do because that's an interface we could document and provide to users.

Reply:
oh, that's not what I meant. I thought you were going to rewrite the existing mapper to match the new format, so we could get rid of the intermediate wrapper object.

Reply:
if we are not going to use the mapper, we should probably remove it altogether to avoid confusion or accidentally using it.

Reply:
Rewrite the existing mapper to map to what? If not Schema then it has to be some other type that describes the schema. Do you mean just use a generic representation (e.g. Maps, Lists, String) ?